pkg/gcp/ocpgwdeployer.go

```
@@ -0,0 +1,285 @@
+	if err := d.gcp.openPorts(externalIngress); err != nil {
+		reporter.Failed(err)
+		return err
+	}
+
+	reporter.Succeeded("Opened External ports %q with firewall rule %q on GCP",
+		formatPorts(input.PublicPorts), externalIngress.Name)
+
+	reporter.Started(messageRetrieveZones)
+
+	zones, err := d.gcp.client.ListZones()
+	if err != nil {
+		reporter.Failed(err)
+		return fmt.Errorf("failed to list the zones in the project %q. %v", d.gcp.projectID, err)
+	}
+
+	reporter.Succeeded(messageRetrievedZones)
+
+	reporter.Started(messageValidateCurrentGWCount)
+
+	for _, zone := range zones.Items {
+		if zone.Region != d.gcp.region {
+			continue
+		}
+
+		instanceList, err := d.gcp.client.ListInstances(zone.Name)
+		if err != nil {
+			reporter.Failed(err)
+			return fmt.Errorf("failed to list instances in zone %q of project project %q. %v", zone.Name, d.gcp.projectID, err)
+		}
+
+		for _, instance := range instanceList.Items {
+			hasPublicIP, err := d.gcp.client.InstanceHasPublicIP(instance)
+			if err != nil {
+				reporter.Failed(err)
+				return fmt.Errorf("failed to verify if instance %q has public-ip or not in project %q. %v", instance.Name, d.gcp.projectID, err)
+			}
+
+			if hasPublicIP {
+				for _, tag := range instance.Tags.Items {
+					if tag == submarinerGatewayNodeTag {
+						currentGWInstanceList = append(currentGWInstanceList, instance)
+						break
+					}
+				}
+			}
+		}
+	}
+
+	if len(currentGWInstanceList) == input.Gateways {
+		reporter.Succeeded(messageValidatedCurrentGWs)
+		return nil
+	}
+
+	if len(currentGWInstanceList) < input.Gateways {
+		gatewayNodesToDeploy := input.Gateways - len(currentGWInstanceList)
+
+		reporter.Started(messageDeployGatewayNode)
+
+		for _, zone := range zones.Items {
+			err := d.deployGateway(zone.Name)
+			if err != nil {
+				reporter.Failed(err)
+				return err
+			}
+
+			gatewayNodesToDeploy--
+			if gatewayNodesToDeploy <= 0 {
+				reporter.Succeeded(messageDeployedGatewayNode)
+				return nil
+			}
+		}
```

So what if `gatewayNodesToDeploy` doesn't reach 0 and we end up here. That would mean we didn't find enough zones to satisfy the requested # of gateways. Should we return an error and/or report it?

---

Added a check